### ISSUE TITLE
When torch.compile is used, some APIs are decomposed even if decomposition is empty.

### ISSUE THREAD

**Issue by wangjiangben-hw** [pytorch/pytorch#112744]
### 🐛 Describe the bug

When torch.compile is used, some APIs are decomposed even if decomposition is empty.

For example, https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py#L2456 
When torch.compile upsample_nearest2d is automatically decomposed into aten._unsafe_index and contiguous, which does not meet user expectations.

For a third-party device, for example, if the key of PrivateUsr1, the device uses AutoGrad and then redispatch to AutoGradPrivateUsr1. However, the device is decomposed when compiling upsample_nearest2d, which does not meet the expectation.

I feel that this automatic disassembly is unreasonable, especially the AutoGrad key, and I would like to discuss whether it can be deleted.

Please take a look at it. Is this reasonable?

### Versions

pytorch 2.1

**Comment by Chillee** [comment]
This is because `upsample_nearest2d` is decomposed at the "compositeimplicitautograd" level, and we don't currently have a way to disable that in `torch.compile`.

cc: @ezyang @SherlockNoMad

**Comment by wangjiangben-hw** [comment]
I would suggest that it might be better to remove the py_impl registration of the AutoGrad Key first, which is enabled by default in decompose.

Compositeimplicitautograd is easy to give the user an illusion. I hope to obtain the complete semantic expression of the aten.upsample_nearest2d.default operator on the fx graph. If it is implicitly decomposed, on the one hand, the user does not get the desired semantic information. On the other hand, it may not be an efficient choice on different devices (e.g. XPU, NPU, IPU, etc.).

Please help to have a look, 3q~

**Comment by zhxchen17** [comment]
Hi @wangjiangben-hw, just FYI if you're using torch.export, we will return a graph at "pre dispatch" level which might preserve the op you want to keep, also we will give knobs to user to have some control over the decomposition behavior. 

Do you still need this feature? Do you mind sharing a bit more on your use case and why which part of your code needs to consume upsample_nearest2d op as a whole? Thanks!

**Comment by zou3519** [comment]
> I would suggest that it might be better to remove the py_impl registration of the AutoGrad/CompositeImplicitAutograd Key first, which is enabled by default in decomposition_table(https://github.com/pytorch/pytorch/blob/main/torch/_decomp/__init__.py#L29).

This is exactly what we're doing in export today

**Comment by wangjiangben-hw** [comment]
> Hi @wangjiangben-hw, just FYI if you're using torch.export, we will return a graph at "pre dispatch" level which might preserve the op you want to keep, also we will give knobs to user to have some control over the decomposition behavior. cc @tugsbayasgalan
> 
> Do you still need this feature? Do you mind sharing a bit more on your use case and which part of your code needs to consume upsample_nearest2d op as a whole? Thanks!

Yes, we still need this feature. We have other compilation backends on our side. Higher-level ops can express more semantic information for us, which helps our compiler further.
